inginious/frontend/plugins/upcoming_tasks/__init__.py

```diff
@@ -0,0 +1,153 @@
+                        if self.user_manager.course_is_open_to_user(course, username, False) and
+                        self.user_manager.course_is_user_registered(course, username)}
+        open_courses = OrderedDict(sorted(iter(open_courses.items()), key=lambda x: x[1].get_name(self.user_manager.session_language())))
+
+        courses = {self.course_factory.get_course(courseid): course for courseid, course in all_courses.items()
+                   if self.user_manager.course_is_open_to_user(course, username, False) and
+                   self.user_manager.course_is_user_registered(course, username)}
+
+        """ Get last submissions for left pannel """
+        last_submissions = self.submission_manager.get_user_last_submissions(5, {"courseid": {"$in": list(open_courses.keys())}})
+        except_free_last_submissions = []
+        for submission in last_submissions:
+            try:
+                submission["task"] = open_courses[submission['courseid']].get_task(submission['taskid'])
+                except_free_last_submissions.append(submission)
+            except:
+                pass
+
+        """ Get the courses tasks, remove finished ones and courses that have no available unfinished tasks with upcoming deadline in range """
+        tasks_data = {}
+        outdated_tasks=[]
+        for course in courses:
+            tasks = course.get_tasks()
+            for task in tasks:
+                the_task = course.get_task(task)
+                if ( (not the_task.get_accessible_time().is_open()) or ( (the_task.get_accessible_time().get_soft_end_date()) > (datetime.now()+timedelta(days=time_planner)) )): 
+                    outdated_tasks += [task]
+            new_user_task_list = course.get_task_dispenser().get_user_task_list([username])[username]
+            tasks_data.update({taskid: {"succeeded": False, "grade": 0.0} for taskid in new_user_task_list})
+            user_tasks = self.database.user_tasks.find({"username": username, "courseid": course.get_id(), "taskid": {"$in": new_user_task_list}})
+            for user_task in user_tasks:
+                tasks_data[user_task["taskid"]]["succeeded"] = user_task["succeeded"]
+                tasks_data[user_task["taskid"]]["grade"] = user_task["grade"]
+                if ( tasks_data[user_task["taskid"]]["succeeded"]): 
+                    del tasks_data[user_task["taskid"]]
+            """ Remove outdated tasks and LTi courses or courses with no unfinished available tasks with deadline """
+            for outdated_task in outdated_tasks:
+                if outdated_task in tasks_data:
+                    del tasks_data[outdated_task]
+            if ((not any(task in tasks for task in tasks_data)) or course.is_lti()):
+                del open_courses[course.get_id()]
+        my_render=RenderOrdered(username)
+        time_planner = ["7", "14", "30", "unlimited"]
+        tasks_data_keys = list(tasks_data.keys())
```

Comment:
What happens if you have tasks with same taskid in different courses ?

Reply:
Bad things...
Currently resolving this issue.

Reply:
Now tasks are filtered based on courseid and taskid.
Having multiple courses using same tasks id is not an issue anymore.